File: packages/python/src/kymo/icons_cli.py
"""`kymo icons <verb>` — the icon command group (CR-ICONS-001 / FR-12..15).

Verbs over the Icons v2 catalogue (P3 per-set IconifyJSON):

    kymo icons list     [provider] [--json]
    kymo icons search   <query> [--provider P] [--remote] [--limit N] [--json]
    kymo icons describe <prefix:name> [--json]
    kymo icons download <prefix:name>... [--from <source>] [-o <dir>] [-y]

`list`/`search`/`describe` are offline (no network); only `search --remote`
and `download --from iconify` touch the Iconify API (stdlib `urllib`, so
`packages/js` stays zero-dep on its side too). Every verb exits 0 on success
and non-zero on error (unknown key, malformed address, network failure) so the
group is CI-usable. `icons` is the only reserved first token — every other
first token is a converter source path (the verb-less converter is unchanged).
"""
from __future__ import annotations

import json as _json
import sys

from . import icons

_USAGE = (
    "usage: kymo icons <list|search|describe|download> [args]\n"
    "  list     [provider] [--json]\n"
    "  search   <query> [--provider P] [--remote] [--limit N] [--json]\n"
    "  describe <prefix:name> [--json]\n"
    "  download <prefix:name>... [--from <source>] [-o <dir>] [-y]"
)
_ICONIFY_API = "https://api.iconify.design"


def _opt(args: list[str], name: str) -> str | None:
    """Pop `--name value` (or `-o value`) from `args`, returning the value."""
    if name in args:
        i = args.index(name)
        if i + 1 < len(args):
            val = args[i + 1]
            del args[i:i + 2]
            return val
        del args[i]
    return None


def _err(msg: str) -> int:
    print(msg, file=sys.stderr)
    return 1


# ── list (FR-13) ──────────────────────────────────────────────────────────
def _list(args: list[str]) -> int:
    as_json = "--json" in args
    rest = [a for a in args if a != "--json"]
    cols = icons.collections()
    if not rest:
        if as_json:
            print(_json.dumps(cols, ensure_ascii=False))
            return 0
        if not cols:
            print("no icon sets found (run the generator: npm run build-manifest)")
            return 0
        for prefix in sorted(cols):
            c = cols[prefix]
            print(f"{prefix:<16} {c['total']:>5} icons  [{', '.join(c['categories'])}]")
        return 0
    prefix = rest[0]
    s = icons.load_set(prefix)
    if not s:
        return _err(f"unknown provider: {prefix!r}")
    addrs = [f"{prefix}:{name}" for name in sorted(s["icons"])]
    if as_json:
        print(_json.dumps(addrs, ensure_ascii=False))
    else:
        for a in addrs:
            print(a)
    return 0


# ── search (FR-14) ──────────────────────────────────────────────────────────
def _search(args: list[str]) -> int:
    as_json = "--json" in args
    remote = "--remote" in args
    provider = _opt(args, "--provider")
    limit = _opt(args, "--limit")
    rest = [a for a in args if a not in ("--json", "--remote")]
    if not rest:
        return _err("usage: kymo icons search <query> [--provider P] [--remote] [--limit N] [--json]")
    query = rest[0].lower()
    n = int(limit) if limit and limit.isdigit() else 50

    prefixes = [provider] if provider else sorted(icons.collections())
    scored: list[tuple[int, str]] = []
    for prefix in prefixes:
        s = icons.load_set(prefix)
        for name, rec in s.get("icons", {}).items():
            addr = f"{prefix}:{name}"
            hay_name = name.lower()
            cat = (rec.get("category") or "").lower()
            if query == hay_name:
                rank = 0
            elif query in hay_name:
                rank = 1
            elif query in cat:
                rank = 2
            else:
                continue
            scored.append((rank, addr))
    scored.sort(key=lambda t: (t[0], t[1]))
    results = [addr for _, addr in scored][:n]

    remote_results: list[str] = []
    if remote:
        try:
            remote_results = _iconify_search(query, n)
        except Exception as exc:                      # network failure → non-zero
            return _err(f"remote search failed: {exc}")

    if as_json:
        print(_json.dumps(
            {"query": query, "results": results, "remote": remote_results},
            ensure_ascii=False,
        ))
        return 0
    for a in results:
        print(a)
    for a in remote_results:
        print(f"{a}\t(iconify, fetchable)")
    return 0


def _iconify_search(query: str, limit: int) -> list[str]:
    from urllib.parse import urlencode
    from urllib.request import urlopen

    url = f"{_ICONIFY_API}/search?" + urlencode({"query": query, "limit": limit})
    with urlopen(url, timeout=10) as resp:            # noqa: S310 (https only)
        data = _json.loads(resp.read().decode("utf-8"))
    return list(data.get("icons", []))


# ── describe (FR-15) ──────────────────────────────────────────────────────
def _describe(args: list[str]) -> int:
    as_json = "--json" in args
    rest = [a for a in args if a != "--json"]
    if not rest:
        return _err("usage: kymo icons describe <prefix:name> [--json]")
    addr = rest[0]
    if not icons.is_address(addr):
        return _err(f"malformed address (expected prefix:name): {addr!r}")
    prefix, _, name = addr.partition(":")
    s = icons.load_set(prefix)
    rec = s.get("icons", {}).get(name) if s else None
    if rec is None:
        return _err(f"unknown icon: {addr!r}")
    info = s.get("info", {})
    out = {
        "address": addr,
        "width": rec.get("width", s.get("width")),
        "height": rec.get("height", s.get("height")),
        "category": rec.get("category"),
        "path": rec.get("path"),
        "set": {"name": info.get("name", prefix), "total": info.get("total"),
                "license": info.get("license")},
        "aliasChain": _alias_chain(addr),
    }
    if as_json:
        print(_json.dumps(out, ensure_ascii=False))
        return 0
    print(f"address  : {out['address']}")
    print(f"size     : {out['width']}×{out['height']}")
    print(f"category : {out['category']}")
    print(f"source   : {out['path']}")
    print(f"set      : {out['set']['name']} ({out['set']['total']} icons)")
    if out["aliasChain"]:
        print(f"alias of : {' → '.join(out['aliasChain'])}")
    return 0


def _alias_chain(addr: str) -> list[str]:
    chain: list[str] = []
    seen: set[str] = set()
    cur = addr
    while cur in icons._ALIASES and cur not in seen:
        seen.add(cur)
        cur = icons._ALIASES[cur]["parent"]
        chain.append(cur)
    return chain


# ── download (FR-15) ──────────────────────────────────────────────────────
def _download(args: list[str]) -> int:
    source = _opt(args, "--from")
    out_dir = _opt(args, "-o") or "icons"
    yes = "-y" in args
    targets = [a for a in args if a not in ("-y",)]
    if not targets:
        return _err("usage: kymo icons download <prefix:name>... [--from <source>] [-o <dir>] [-y]")

    from pathlib import Path
    dest = Path(out_dir)
    for addr in targets:
        if not icons.is_address(addr):
            return _err(f"malformed address: {addr!r}")
        if source == "iconify":
            try:
                body = _iconify_fetch(addr)
            except Exception as exc:
                return _err(f"iconify fetch failed for {addr}: {exc}")
            try:
                from .icons_pipeline import normalize  # P4 normalize (FR-8)
                body = normalize(body)
            except Exception:
                pass                                           # pipeline optional pre-P4
            prefix, _, name = addr.partition(":")
            target = dest / prefix / f"{name}.svg"
            if target.exists() and not yes:
                return _err(f"refusing to overwrite {target} (pass -y)")
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_text(body, encoding="utf-8")
            print(f"✓ wrote {target}")
        else:
            # Local copy from the vendored catalogue.
            src = icons._NS_ICONS.get(addr)
            if src is None:
                return _err(f"unknown icon: {addr!r}")
            target = dest / src.name
            if target.exists() and not yes:
                return _err(f"refusing to overwrite {target} (pass -y)")
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_bytes(src.read_bytes())
            print(f"✓ wrote {target}")
    return 0


def _iconify_fetch(addr: str) -> str:
    from urllib.request import urlopen

    prefix, _, name = addr.partition(":")
    url = f"{_ICONIFY_API}/{prefix}/{name}.svg"
    with urlopen(url, timeout=10) as resp:            # noqa: S310 (https only)
        return resp.read().decode("utf-8")


# ── dispatch ────────────────────────────────────────────────────────────────
def run(argv: list[str]) -> int:
    if not argv or argv[0] in ("-h", "--help"):
        print(_USAGE)
        return 0 if argv else 1
    verb, rest = argv[0], list(argv[1:])
    handlers = {"list": _list, "search": _search, "describe": _describe, "download": _download}
    handler = handlers.get(verb)
    if handler is None:
        print(f"unknown icons command: {verb!r}\n{_USAGE}", file=sys.stderr)
        return 2
    return handler(rest)

File: packages/js/bin/icons-cli.mjs
/**
 * `kymo icons <verb>` — JS mirror of packages/python/src/kymo/icons_cli.py
 * (CR-ICONS-001 / FR-12..15). Hand-rolled parser, ZERO runtime dependencies
 * (NFR-3): reads the generated artifacts with `node:fs`, and uses the built-in
 * `fetch` for the opt-in `--remote` / `--from iconify` paths only.
 *
 *   kymo icons list     [provider] [--json]
 *   kymo icons search   <query> [--provider P] [--remote] [--limit N] [--json]
 *   kymo icons describe <prefix:name> [--json]
 *   kymo icons download <prefix:name>... [--from <source>] [-o <dir>] [-y]
 *
 * Every verb returns 0 on success, non-zero on error (CI-usable).
 */
import { readFileSync, writeFileSync, mkdirSync, existsSync, copyFileSync } from "node:fs";
import { join, dirname, basename } from "node:path";
import { fileURLToPath } from "node:url";

const PKG = join(dirname(fileURLToPath(import.meta.url)), "..");   // packages/js
const REPO = join(PKG, "..", "..");
const ICONIFY_API = "https://api.iconify.design";
const ADDR_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*:[a-z0-9]+(?:-[a-z0-9]+)*$/;

const USAGE = [
  "usage: kymo icons <list|search|describe|download> [args]",
  "  list     [provider] [--json]",
  "  search   <query> [--provider P] [--remote] [--limit N] [--json]",
  "  describe <prefix:name> [--json]",
  "  download <prefix:name>... [--from <source>] [-o <dir>] [-y]",
].join("\n");

function readJSON(path, fallback) {
  try { return JSON.parse(readFileSync(path, "utf-8")); }
  catch { return fallback; }
}
const collections = () => readJSON(join(PKG, "icons-collections.json"), {});
const loadSet = (prefix) => readJSON(join(PKG, "sets", `${prefix}.json`), null);
const manifest = () => readJSON(join(PKG, "icons-manifest.json"), { icons: {}, legacy: {}, aliases: {} });

/** Pop `--name value` (or `-o value`) from argv, returning the value. */
function opt(argv, name) {
  const i = argv.indexOf(name);
  if (i >= 0 && i + 1 < argv.length) { const v = argv[i + 1]; argv.splice(i, 2); return v; }
  if (i >= 0) { argv.splice(i, 1); }
  return undefined;
}

function aliasChain(addr) {
  const aliases = manifest().aliases ?? {};
  const chain = []; const seen = new Set(); let cur = addr;
  while (aliases[cur] && !seen.has(cur)) { seen.add(cur); cur = aliases[cur].parent; chain.push(cur); }
  return chain;
}

// ── list (FR-13) ───────────────────────────────────────────────────────────
function cmdList(argv, log, err) {
  const asJson = argv.includes("--json");
  const rest = argv.filter((a) => a !== "--json");
  const cols = collections();
  if (rest.length === 0) {
    if (asJson) { log(JSON.stringify(cols)); return 0; }
    if (!Object.keys(cols).length) { log("no icon sets found (run: npm run build-manifest)"); return 0; }
    for (const prefix of Object.keys(cols).sort()) {
      const c = cols[prefix];
      log(`${prefix.padEnd(16)} ${String(c.total).padStart(5)} icons  [${c.categories.join(", ")}]`);
    }
    return 0;
  }
  const prefix = rest[0];
  const s = loadSet(prefix);
  if (!s) { err(`unknown provider: ${JSON.stringify(prefix)}`); return 1; }
  const addrs = Object.keys(s.icons).sort().map((n) => `${prefix}:${n}`);
  log(asJson ? JSON.stringify(addrs) : addrs.join("\n"));
  return 0;
}

// ── search (FR-14) ───────────────────────────────────────────────────────────
async function cmdSearch(argv, log, err) {
  const asJson = argv.includes("--json");
  const remote = argv.includes("--remote");
  const provider = opt(argv, "--provider");
  const limit = opt(argv, "--limit");
  const rest = argv.filter((a) => a !== "--json" && a !== "--remote");
  if (!rest.length) { err("usage: kymo icons search <query> [--provider P] [--remote] [--limit N] [--json]"); return 1; }
  const query = rest[0].toLowerCase();
  const n = limit && /^\d+$/.test(limit) ? Number(limit) : 50;

  const prefixes = provider ? [provider] : Object.keys(collections()).sort();
  const scored = [];
  for (const prefix of prefixes) {
    const s = loadSet(prefix);
    if (!s) continue;
    for (const [name, rec] of Object.entries(s.icons)) {
      const cat = (rec.category ?? "").toLowerCase();
      let rank;
      if (query === name) rank = 0;
      else if (name.includes(query)) rank = 1;
      else if (cat.includes(query)) rank = 2;
      else continue;
      scored.push([rank, `${prefix}:${name}`]);
    }
  }
  scored.sort((a, b) => (a[0] - b[0]) || (a[1] < b[1] ? -1 : 1));
  const results = scored.map(([, a]) => a).slice(0, n);

  let remoteResults = [];
  if (remote) {
    try {
      const r = await fetch(`${ICONIFY_API}/search?query=${encodeURIComponent(query)}&limit=${n}`);
      if (!r.ok) throw new Error(`HTTP ${r.status}`);
      remoteResults = (await r.json()).icons ?? [];
    } catch (e) { err(`remote search failed: ${e.message}`); return 1; }
  }
  if (asJson) { log(JSON.stringify({ query, results, remote: remoteResults })); return 0; }
  for (const a of results) log(a);
  for (const a of remoteResults) log(`${a}\t(iconify, fetchable)`);
  return 0;
}

// ── describe (FR-15) ──────────────────────────────────────────────────────────
function cmdDescribe(argv, log, err) {
  const asJson = argv.includes("--json");
  const rest = argv.filter((a) => a !== "--json");
  if (!rest.length) { err("usage: kymo icons describe <prefix:name> [--json]"); return 1; }
  const addr = rest[0];
  if (!ADDR_RE.test(addr)) { err(`malformed address (expected prefix:name): ${JSON.stringify(addr)}`); return 1; }
  const [prefix, name] = [addr.slice(0, addr.indexOf(":")), addr.slice(addr.indexOf(":") + 1)];
  const s = loadSet(prefix);
  const rec = s && s.icons[name];
  if (!rec) { err(`unknown icon: ${JSON.stringify(addr)}`); return 1; }
  const info = s.info ?? {};
  const out = {
    address: addr,
    width: rec.width ?? s.width,
    height: rec.height ?? s.height,
    category: rec.category ?? null,
    path: rec.path ?? null,
    set: { name: info.name ?? prefix, total: info.total ?? null, license: info.license ?? null },
    aliasChain: aliasChain(addr),
  };
  if (asJson) { log(JSON.stringify(out)); return 0; }
  log(`address  : ${out.address}`);
  log(`size     : ${out.width}×${out.height}`);
  log(`category : ${out.category}`);
  log(`source   : ${out.path}`);
  log(`set      : ${out.set.name} (${out.set.total} icons)`);
  if (out.aliasChain.length) log(`alias of : ${out.aliasChain.join(" → ")}`);
  return 0;
}

// ── download (FR-15) ──────────────────────────────────────────────────────────
async function cmdDownload(argv, log, err) {
  const source = opt(argv, "--from");
  const outDir = opt(argv, "-o") || "icons";
  const yes = argv.includes("-y");
  const targets = argv.filter((a) => a !== "-y");
  if (!targets.length) { err("usage: kymo icons download <prefix:name>... [--from <source>] [-o <dir>] [-y]"); return 1; }

  for (const addr of targets) {
    if (!ADDR_RE.test(addr)) { err(`malformed address: ${JSON.stringify(addr)}`); return 1; }
    const [prefix, name] = [addr.slice(0, addr.indexOf(":")), addr.slice(addr.indexOf(":") + 1)];
    if (source === "iconify") {
      let body;
      try {
        const r = await fetch(`${ICONIFY_API}/${prefix}/${name}.svg`);
        if (!r.ok) throw new Error(`HTTP ${r.status}`);
        body = await r.text();
      } catch (e) { err(`iconify fetch failed for ${addr}: ${e.message}`); return 1; }
      try {
        const { normalize } = await import("../scripts/icons-pipeline.mjs");   // P4 (FR-8)
        body = normalize(body);
      } catch { /* pipeline optional pre-P4 */ }
      const target = join(outDir, prefix, `${name}.svg`);
      if (existsSync(target) && !yes) { err(`refusing to overwrite ${target} (pass -y)`); return 1; }
      mkdirSync(dirname(target), { recursive: true });
      writeFileSync(target, body);
      log(`✓ wrote ${target}`);
    } else {
      const path = manifest().icons[addr];
      if (!path) { err(`unknown icon: ${JSON.stringify(addr)}`); return 1; }
      const target = join(outDir, basename(path));
      if (existsSync(target) && !yes) { err(`refusing to overwrite ${target} (pass -y)`); return 1; }
      mkdirSync(dirname(target), { recursive: true });
      copyFileSync(join(REPO, path), target);
      log(`✓ wrote ${target}`);
    }
  }
  return 0;
}

const HANDLERS = { list: cmdList, search: cmdSearch, describe: cmdDescribe, download: cmdDownload };

export async function run(argv, log = console.log, err = console.error) {
  if (!argv.length || argv[0] === "-h" || argv[0] === "--help") { log(USAGE); return argv.length ? 0 : 1; }
  const [verb, ...rest] = argv;
  const handler = HANDLERS[verb];
  if (!handler) { err(`unknown icons command: ${JSON.stringify(verb)}\n${USAGE}`); return 2; }
  return handler(rest, log, err);
}
